ably/proto_message.go

```diff
@@ -88,9 +83,10 @@ func (m Message) withEncodedData(cipher channelCipher) (Message, error) {
 	if err != nil {
 		return Message{}, fmt.Errorf("encrypting message data: %w", err)
 	}
+	// if the plain text is a string, then add utf-8 to the list of encodings
+	// to indicate to the client-side decoder that the plain text should be
+	// further decoded into a utf-8 string after being decrypted.
```

Comment:
Correct me if I am wrong, according to @SimonWoolf UTF-8 represents binary payload. So, we shouldn't mention UTF-8 encoding in the first place, because we are sending payload as a string (Encoding doesn't matter, it's a string in the end). I am not sure, but I feel this might case issue on js client, since it will expect binary payload as a part of UTF-8 encoding.

Reply:
Also, not sure `if check` makes sense or not, Since, it's at a point where `_, ok := m.Data.(string)` will always return true

Reply:
> UTF-8 represents binary payload

That is correct, and when the client decrypts this cipher-text, they will get a binary payload, and thus we need to include the utf-8 encoding to further indicate that this decrypted binary payload is in fact a utf-8 encoded string.

This is distinctly different to when the publisher simply publishes a string, because then it's just a string, there is no binary payload involved.

Does that make sense? This should definitely be clarified when addressing the linked docs issue (https://github.com/ably/docs/issues/1165).

Reply:
> not sure if check makes sense or not, Since, it's at a point where _, ok := m.Data.(string) will always return true

Let's keep the check, as it makes the logic explicit that it only applies to string payloads, and does not rely on other logic in the file that may change in the future.

Reply:
Then you need to modify `if` check, since current one will always return true. 
`Data` field of `Message` is currently being mutated, to the point, it will always store encoded string. So, it's difficult to guess original encoding binary/text.

Reply:
Then you need to modify `if` check, since current one will always return true. 
`Data` field of `Message` is currently being mutated, to the point, it will always store encoded string. So, it's difficult to guess original encoding binary/text.

Reply:
Also, there should be one more check whether its a valid `utf-8` string.

Reply:
> Then you need to modify if check, since current one will always return true.

As mentioned in my last comment, I'm going to keep the check, as it defends against logic changing elsewhere and utf-8 encoding being added when `m.Data` is not a string.

> Data field of Message is currently being mutated, to the point, it will always store encoded string. So, it's difficult to guess original encoding binary/text.

Since this encoding logic is iterative, there is no need to know the encoding of the _original_ payload, just whatever the payload of `m.Data` is at the time of adding the next encoding in the chain.

For example, if `m.Data` is originally some arbitrary object, then it will get serialised as JSON and the `json` encoding added [here](https://github.com/ably/ably-go/blob/main/ably/proto_message.go#L73-L74). Now, from this point on, the payload _is_ a utf-8 encoded string, it doesn't matter from that point onwards that the original payload was not a string, since that's been covered by the addition of the `json` encoding.

This is explicitly tested [here](https://github.com/ably/ably-go/blob/main/ably/proto_message_test.go#L76-L86), and this PR should not change those semantics.

> Also, there should be one more check whether its a valid utf-8 string, then only specify UTF-8 encoding.

Good spot, updated in 153c932 👍

Reply:
I am resolving this conversation for the time being. Though I am not sure, if UTF-8 will work or not.